src/EditorFeatures/CSharpTest/UseRecursivePatterns/UseRecursivePatternsTests.cs

```diff
@@ -0,0 +1,515 @@
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        switch (node)
+        {
+            case String
+            {
+                Length: 0
+            }
+
+            s when b:
+                break;
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test16()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    void M()
+    {
+        switch (node)
+        {
+            case String s [||]when b && s.Length is int && c:
+                break;
+        }
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        switch (node)
+        {
+            case String
+            {
+                Length: int _
+            }
+
+            s when b && c:
+                break;
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test17()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    void M()
+    {
+        switch (node)
+        {
+            case String s [||]when s.Length is int && b && c:
+                break;
+        }
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        switch (node)
+        {
+            case String
+            {
+                Length: int _
+            }
+
+            s when b && c:
+                break;
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test18()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    void M()
+    {
+        switch (node)
+        {
+            case String s [||]when b && c && s.Length is int:
+                break;
+        }
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        switch (node)
+        {
+            case String
+            {
+                Length: int _
+            }
+
+            s when b && c:
+                break;
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test11()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return b.P1 == 1 && b.P2 == 2 [||]&& x;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return b is { P1: 1, P2: 2 } && x;
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test12()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return x && b.P1 == 1 [||]&& b.P2 == 2;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return x && b is
+        {
+            P1: 1, P2: 2
+        };
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test13()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return x is var v [||]&& v != null;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return x is { } v ;
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test14()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M(C x)
+    {
+        return x is var v [||]&& v is object;
```

Comment:
In case `is object` could not be converted to a null check we offer the fix anyways, this has two consequence:

- We've changed the definite assignment state of `v`
- We've changed the type of `v`.

The latter point is probably ok since there's a type check anyways but I think we should not offer the fix here if `v` is used anywhere else because it'll be reported as not definitely assigned. Thoughts?